File: forms/palliative-assessment/back-end-with-loco/tests/engine/esas_grader_test.rs
use palliative_assessment_loco_crate::engine::esas_grader::grade;
use palliative_assessment_loco_crate::engine::rules::all_rules;
use palliative_assessment_loco_crate::engine::types::*;

/// A baseline assessment with all ten ESAS-r items set to 0 (none).
fn baseline_case() -> AssessmentData {
    AssessmentData {
        esasr_symptoms: EsasrSymptoms {
            pain: Some(0),
            tiredness: Some(0),
            drowsiness: Some(0),
            nausea: Some(0),
            lack_of_appetite: Some(0),
            shortness_of_breath: Some(0),
            depression: Some(0),
            anxiety: Some(0),
            wellbeing: Some(0),
            other: Some(0),
            other_label: String::new(),
            symptom_notes: String::new(),
        },
        ..AssessmentData::default()
    }
}

#[test]
fn empty_assessment_has_total_zero_and_no_answered_items() {
    let data = AssessmentData::default();
    let result = grade(&data);
    assert_eq!(result.esas_total, 0);
    assert_eq!(result.answered_count, 0);
    assert_eq!(result.severity_band, "none");
}

#[test]
fn baseline_zero_assessment_classifies_as_none() {
    let data = baseline_case();
    let result = grade(&data);
    assert_eq!(result.esas_total, 0);
    assert_eq!(result.answered_count, 10);
    assert_eq!(result.severity_band, "none");
}

#[test]
fn moderate_total_classifies_as_moderate() {
    let mut data = baseline_case();
    data.esasr_symptoms.pain = Some(5);
    data.esasr_symptoms.tiredness = Some(5);
    data.esasr_symptoms.shortness_of_breath = Some(5);
    data.esasr_symptoms.depression = Some(5);
    data.esasr_symptoms.anxiety = Some(5);
    data.esasr_symptoms.wellbeing = Some(6);
    data.esasr_symptoms.other = Some(5);
    let result = grade(&data);
    // 5+5+5+5+5+6+5 + zeroes = 36 -> moderate (31-60)
    assert_eq!(result.esas_total, 36);
    assert_eq!(result.severity_band, "moderate");
}

#[test]
fn severe_pain_yields_individual_flag() {
    let mut data = baseline_case();
    data.esasr_symptoms.pain = Some(9);
    let result = grade(&data);
    assert!(
        result
            .individual_flags
            .iter()
            .any(|f| f.symptom_key == "pain" && f.score == 9),
        "expected individual flag for pain"
    );
}

#[test]
fn rule_ids_are_unique() {
    let rules = all_rules();
    let mut ids: Vec<&str> = rules.iter().map(|r| r.id).collect();
    ids.sort();
    let len_before = ids.len();
    ids.dedup();
    assert_eq!(ids.len(), len_before, "rule IDs must be unique");
}

#[test]
fn ancillary_low_performance_status_fires() {
    let mut data = baseline_case();
    data.performance_status.pps_score = Some(20);
    let result = grade(&data);
    assert!(result.fired_rules.iter().any(|r| r.id == "PALL-ANC-002"));
}

#[test]
fn ancillary_carer_overwhelmed_fires_with_score_two() {
    let mut data = baseline_case();
    data.carer_family_support.carer_strain_level = "overwhelmed".to_string();
    let result = grade(&data);
    let r = result
        .fired_rules
        .iter()
        .find(|r| r.id == "PALL-ANC-005")
        .expect("carer-burden rule");
    assert_eq!(r.score, 2);
}

#[test]
fn unanswered_esas_items_are_excluded_from_total() {
    let mut data = baseline_case();
    // Wipe pain, tiredness; remaining 8 items contribute.
    data.esasr_symptoms.pain = None;
    data.esasr_symptoms.tiredness = None;
    data.esasr_symptoms.depression = Some(4);
    let result = grade(&data);
    assert_eq!(result.answered_count, 8);
    assert_eq!(result.esas_total, 4);
}

File: forms/palliative-assessment/front-end-with-svelte/src/lib/engine/esas-grader.test.ts
import { describe, it, expect } from 'vitest';
import { calculatePalliativeGrade, gradeESAS } from './esas-grader';
import { detectAdditionalFlags } from './flagged-issues';
import { rules } from './rules';
import { classifyESASTotal, classifyIndividualSymptom } from './utils';
import type { AssessmentData } from './types';

// A fully-blank assessment, built inline so the pure-engine test stays
// independent of the Svelte runes store (which cannot run under plain Vitest).
function blankAssessment(): AssessmentData {
	return {
		demographics: {
			firstName: '',
			lastName: '',
			dateOfBirth: '',
			sex: '',
			nhsOrMrnNumber: '',
			preferredLanguage: '',
			ethnicity: '',
			reporterRole: '',
			reporterName: '',
			assessmentDate: '',
			assessmentSetting: ''
		},
		primaryDiagnosisPrognosis: {
			primaryDiagnosis: '',
			secondaryDiagnoses: '',
			dateOfDiagnosis: '',
			stageOrSeverity: '',
			diseaseProgressing: '',
			prognosisHorizon: '',
			surpriseQuestion: '',
			prognosticIndicators: '',
			relevantTreatmentHistory: ''
		},
		esasrSymptoms: {
			pain: null,
			tiredness: null,
			drowsiness: null,
			nausea: null,
			lackOfAppetite: null,
			shortnessOfBreath: null,
			depression: null,
			anxiety: null,
			wellbeing: null,
			other: null,
			otherLabel: '',
			symptomNotes: ''
		},
		performanceStatus: {
			ppsScore: null,
			akpsScore: null,
			ecogScore: null,
			mobilityNotes: '',
			activityLevel: '',
			bedBound: '',
			requiresAssistanceWithAdls: '',
			adlNotes: ''
		},
		goalsOfCareACP: {
			patientPrioritiesAndWishes: '',
			preferredPlaceOfCare: '',
			preferredPlaceOfDeath: '',
			respectFormCompleted: '',
			respectFormDate: '',
			adrtCompleted: '',
			adrtDate: '',
			lpaHealthAndWelfare: '',
			lpaName: '',
			dnacprDocumented: '',
			dnacprDate: '',
			ceilingOfTreatmentDiscussed: '',
			ceilingOfTreatmentNotes: ''
		},
		medicationsSymptomControl: {
			regularMedications: [],
			asNeededMedications: [],
			syringeDriverInUse: '',
			syringeDriverDetails: '',
			anticipatoryMedsPrescribed: '',
			anticipatoryMedsNotes: '',
			symptomControlOverall: '',
			barriersToControl: '',
			planNotes: ''
		},
		psychosocialSpiritualConcerns: {
			moodConcerns: '',
			moodNotes: '',
			anxietyConcerns: '',
			anxietyNotes: '',
			existentialDistress: '',
			existentialNotes: '',
			spiritualSupportRequested: '',
			faithOrBelief: '',
			chaplaincyNotes: '',
			unresolvedConcerns: '',
			unresolvedNotes: ''
		},
		carerFamilySupport: {
			primaryCarerName: '',
			primaryCarerRelationship: '',
			carerLivesWithPatient: '',
			carerStrainReported: '',
			carerStrainLevel: '',
			carerStrainNotes: '',
			respiteRequired: '',
			respiteNotes: '',
			childrenInHousehold: '',
			childrenSupportNotes: '',
			bereavementRiskIdentified: '',
			bereavementNotes: ''
		},
		multidisciplinaryPlan: {
			specialistPalliativeCareInvolved: '',
			communityNursingInvolved: '',
			hospiceReferralMade: '',
			socialWorkReferralMade: '',
			occupationalTherapyReferralMade: '',
			physiotherapyReferralMade: '',
			dieticianReferralMade: '',
			chaplaincyReferralMade: '',
			psychologyReferralMade: '',
			otherReferrals: '',
			reviewInterval: '',
			keyWorkerName: '',
			planSummary: ''
		}
	};
}

function withSymptoms(scores: Partial<AssessmentData['esasrSymptoms']>): AssessmentData {
	const d = blankAssessment();
	d.esasrSymptoms = { ...d.esasrSymptoms, ...scores };
	return d;
}

describe('classifyESASTotal', () => {
	it('bands the total 0-100 correctly', () => {
		expect(classifyESASTotal(0)).toBe('none');
		expect(classifyESASTotal(10)).toBe('none');
		expect(classifyESASTotal(11)).toBe('mild');
		expect(classifyESASTotal(30)).toBe('mild');
		expect(classifyESASTotal(31)).toBe('moderate');
		expect(classifyESASTotal(60)).toBe('moderate');
		expect(classifyESASTotal(61)).toBe('severe');
		expect(classifyESASTotal(100)).toBe('severe');
	});
});

describe('classifyIndividualSymptom', () => {
	it('bands a single symptom 0-10', () => {
		expect(classifyIndividualSymptom(null)).toBe('');
		expect(classifyIndividualSymptom(0)).toBe('none');
		expect(classifyIndividualSymptom(3)).toBe('mild');
		expect(classifyIndividualSymptom(6)).toBe('moderate');
		expect(classifyIndividualSymptom(7)).toBe('severe');
		expect(classifyIndividualSymptom(10)).toBe('severe');
	});
});

describe('ESAS-r grading engine', () => {
	it('returns a blank/none grade for an unanswered assessment', () => {
		const result = calculatePalliativeGrade(blankAssessment());
		expect(result.esasTotal).toBe(0);
		expect(result.answeredCount).toBe(0);
		expect(result.severityBand).toBe('none');
		expect(result.individualFlags).toHaveLength(0);
		expect(result.firedRules).toHaveLength(0);
	});

	it('sums only answered symptoms and excludes nulls from the count', () => {
		const result = gradeESAS(withSymptoms({ pain: 5, tiredness: 3, nausea: null }));
		expect(result.esasTotal).toBe(8);
		expect(result.answeredCount).toBe(2);
	});

	it('classifies a severe total (61-100)', () => {
		const result = calculatePalliativeGrade(
			withSymptoms({
				pain: 8,
				tiredness: 8,
				drowsiness: 7,
				nausea: 7,
				lackOfAppetite: 8,
				shortnessOfBreath: 7,
				depression: 7,
				anxiety: 7,
				wellbeing: 8,
				other: 7
			})
		);
		expect(result.esasTotal).toBe(74);
		expect(result.severityBand).toBe('severe');
	});

	it('raises an individual flag for any symptom >= 7', () => {
		const result = gradeESAS(withSymptoms({ pain: 8, tiredness: 2 }));
		expect(result.individualFlags.some((f) => f.symptomKey === 'pain')).toBe(true);
		expect(result.individualFlags.some((f) => f.symptomKey === 'tiredness')).toBe(false);
	});

	it('fires the uncontrolled-severe-symptom ancillary rule', () => {
		const result = gradeESAS(withSymptoms({ shortnessOfBreath: 9 }));
		expect(result.firedRules.some((r) => r.id === 'PALL-ANC-001')).toBe(true);
	});

	it('has unique rule IDs', () => {
		const ids = rules.map((r) => r.id);
		expect(new Set(ids).size).toBe(ids.length);
	});
});

describe('Palliative flagged-issue detection', () => {
	it('returns no flags for an unanswered assessment', () => {
		expect(detectAdditionalFlags(blankAssessment())).toHaveLength(0);
	});

	it('flags uncontrolled severe pain (high priority)', () => {
		const flags = detectAdditionalFlags(withSymptoms({ pain: 9 }));
		expect(flags.some((f) => f.id === 'FLAG-PALL-PAIN' && f.priority === 'high')).toBe(true);
	});

	it('flags no DNACPR for an end-of-life patient', () => {
		const data = blankAssessment();
		data.performanceStatus.ppsScore = 20;
		data.goalsOfCareACP.dnacprDocumented = 'no';
		const flags = detectAdditionalFlags(data);
		expect(flags.some((f) => f.id === 'FLAG-PALL-DNACPR')).toBe(true);
	});

	it('flags carer overwhelm', () => {
		const data = blankAssessment();
		data.carerFamilySupport.carerStrainLevel = 'overwhelmed';
		const flags = detectAdditionalFlags(data);
		expect(flags.some((f) => f.id === 'FLAG-PALL-CARER-OVERWHELM')).toBe(true);
	});

	it('sorts flags high → medium → low', () => {
		const flags = detectAdditionalFlags(withSymptoms({ pain: 9, tiredness: 5, drowsiness: 2 }));
		const order = { high: 0, medium: 1, low: 2 };
		const priorities = flags.map((f) => f.priority);
		const sorted = [...priorities].sort((a, b) => order[a] - order[b]);
		expect(priorities).toEqual(sorted);
	});
});
